src/fal/packages/environments/base.py

```diff
@@ -176,3 +178,71 @@ def poll_until_result(
                     kind="error",
                 )
                 raise exception
+
+
+@dataclass
+class DualPythonIPC(IsolatedProcessConnection[BaseEnvironment]):
+    # We manage user-defined dual-Python environments in two steps.
+    #   1. Create a primary environment which contains the default dependencies
+    #      to run a fal script (like fal, dbt-core, and required dbt adapters).
+    #   2. Create a secondary environment which contains the user-defined dependencies.
+    #
+    # This is an optimization we apply to reduce the cost of user-defined environments
+    # where we can actually share the primary environment and save a lot of time from not
+    # installing heavy dependencies like dbt adapters again and again. This also heavily
+    # reduces the disk usage.
+
+    primary_path: Path
+    secondary_path: Path
+
+    def start_process(
+        self,
+        service: bridge.Listener,
+        *args,
+        **kwargs,
+    ) -> ContextManager[subprocess.Popen]:
+        # We are going to use the primary environment to run the Python
+        # interpreter, but at the same time we are going to inherit all
+        # the packages from the secondary environment (user's environment)
+        # so that they can technically override anything.
+
+        # The search order is important, we want the secondary path to
+        # take precedence.
+        python_path = python_path_for(self.secondary_path, self.primary_path)
+
+        # The environment which the python executable is going to be used
+        # shouldn't matter much for the virtual-env based installations but
+        # in conda, the Python executable between the primary and the secondary
+        # environment might differ and we'll give precedence to the user's
+        # choice (the secondary environment).
+        python_executable = get_executable_path(self.secondary_path, "python")
+        return subprocess.Popen(
+            [
+                python_executable,
+                isolated_runner.__file__,
+                bridge.encode_service_address(service.address),
+            ],
+            env={"PYTHONPATH": python_path, **os.environ},
+        )
+
+
+def python_path_for(*search_paths) -> str:
+    assert len(search_paths) >= 1
+    return os.pathsep.join(
+        # sysconfig takes the virtual environment path and returns
+        # the directory where all the site packages are located.
+        sysconfig.get_path("purelib", vars={"base": search_path})
```

Comment:
ATL: what is purelib?

Reply:
`purelib` points out to the path where the actual packages are located. E.g. my venv is under `~/.cache/fal/xxx` but the packages actually reside in `~/.cache/fal/xxx/lib/site-packages/something` (depending on the platform), using `purelib` with the base of our search_path (which is the venv directory, or conda) we can find the proper path on where all the site-packages are located.

Reply:
Definitions are encoded in here (and packaging tools like pip/conda use them) https://github.com/python/cpython/blob/c112489f877a37cd6e887f0893fc1f4843c168fa/Lib/sysconfig.py#L88-L118